FirebaseInstallations/Source/Library/FIRSecureStorage.m

```diff
@@ -26,28 +26,108 @@
 #import "FIRInstallationsErrorUtil.h"
 
 @interface FIRSecureStorage ()
-@property(nonatomic, strong) dispatch_queue_t keychainQueue;
+@property(nonatomic, readonly) dispatch_queue_t keychainQueue;
+@property(nonatomic, readonly) dispatch_queue_t inMemoryCacheQueue;
 @property(nonatomic, readonly) NSString *service;
+@property(nonatomic, readonly) NSCache<NSString *, id<NSSecureCoding>> *inMemoryCache;
 @end
 
 @implementation FIRSecureStorage
 
 - (instancetype)init {
-  return [self initWithService:@"com.firebase.FIRInstallations.installations"];
+  NSCache *cache = [[NSCache alloc] init];
+  // Cache up to 5 installations.
+  cache.countLimit = 5;
+  return [self initWithService:@"com.firebase.FIRInstallations.installations" cache:cache];
 }
 
-- (instancetype)initWithService:(NSString *)service {
+- (instancetype)initWithService:(NSString *)service cache:(NSCache *)cache {
   self = [super init];
   if (self) {
-    _keychainQueue = dispatch_queue_create("com.firebase.FIRSecureStorage", DISPATCH_QUEUE_SERIAL);
+    _keychainQueue =
+        dispatch_queue_create("com.firebase.FIRSecureStorage.Keychain", DISPATCH_QUEUE_SERIAL);
+    _inMemoryCacheQueue = dispatch_queue_create("com.firebase.FIRSecureStorage.InMemoryChache",
+                                                DISPATCH_QUEUE_SERIAL);
     _service = [service copy];
+    _inMemoryCache = cache;
   }
   return self;
 }
 
+#pragma mark - Public
+
 - (FBLPromise<id<NSSecureCoding>> *)getObjectForKey:(NSString *)key
                                         objectClass:(Class)objectClass
                                         accessGroup:(nullable NSString *)accessGroup {
+  return [FBLPromise onQueue:self.inMemoryCacheQueue
+                          do:^id _Nullable {
+                            // Return cached object or fail otherwise.
+                            id object = [self.inMemoryCache objectForKey:key];
```

Comment:
Try use ivar as much as possible?
_inMemoryCache?

Reply:
Hm... I usually use ivars in the initializers and getters/setters only to be on the safe side with the memory management and thread safety. Why using ivar is better here?

Reply:
since we already handling the thread safe ourselves by putting them in a queue, I was told it's better to use ivar for thread safe operations.